H1 FROST function, BIP034 hashing function

pdyraga opened the pull request:
~~Depends on #1. Let's merge #1 before we undraft/merge this one.~~

This changeset adds `h1` function. `h1` is the implementation of H1(m) function from [FROST] implemented in a way 
 compatible to how [BIP340] hashing functions are specified. The implementation is copied from @eth-r's prototype. I added tests and annotated relevant parts of the hash functions code with reference to papers. The `testutils` library code was borrowed from the `keep-network/keep-core` repository.

pdyraga committed:
H1 frost function, BIP034 hashing function

This changeset adds H1 function as specified in section 6.5. FROST(secp256k1,
SHA-256) of [FROST]. The [BIP340] hash function is used as expand_message_xmd.
The implementation is copied from @eth-r's prototype. I added tests and
annotated relevant parts of the hash functions code with reference to papers.
The `testutils` library code was borrowed from the `keep-core` repository.

eth-r commented on roast/hash.go:
```diff
@@ -0,0 +1,85 @@
+// [HASH-TO-CURVE] with m=1 as specified in section 6.5. FROST(secp256k1, SHA-256)
+// of [FROST]. The [BIP340] hash function is used as expand_message_xmd.
+func hashToScalar(tag, msg []byte) *big.Int {
+	// The hash_to_field function is specified in [HASH-TO-CURVE] in section
```
We are not actually doing this when implementing BIP-340 which has a much simpler specification for hashing scalars.

pdyraga replied:
👍🏻 Please see https://github.com/threshold-network/roast-go/pull/2/commits/bac833c55b4a5950897ab2193bb854fc37594bb8.

eth-r commented on roast/hash.go:
```diff
@@ -0,0 +1,85 @@
+// The contextString as defined in section 6.5. FROST(secp256k1, SHA-256) of
+// [FROST] is "FROST-secp256k1-SHA256-v1". Since we do a BIP-340 specialized
+// version, we use "FROST-secp256k1-BIP340-v1" as the contextString.
+var contextString = []byte("FROST-secp256k1-SHA256-v1")
```
Mismatch with comment above

pdyraga replied:
Fixed in https://github.com/threshold-network/roast-go/pull/2/commits/7644d8031a9291e45314854d55566cc2402c0a56.

eth-r commented on roast/roast.go:
```diff
@@ -0,0 +1,35 @@
+//	Faz-Hernandez, A., Scott, S., Sullivan, N., Wahby, R. S., and C. A. Wood,
+//	"Hashing to Elliptic Curves", Work in Progress, Internet-Draft,
+//	draft-irtf-cfrg-hash-to-curve- 16, 15 June 2022,
+//	<https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16>.
```
I think HASH-TO-CURVE is just superseded by the BIP-340 hashing specification.

pdyraga replied:
See https://github.com/threshold-network/roast-go/pull/2/commits/bac833c55b4a5950897ab2193bb854fc37594bb8 , please.

eth-r reviewed (changes requested):
A couple of errors regarding the comments and the context string. I don't think a BIP-340 actually uses HASH-TO-CURVE specifications for anything because it has its own, simpler hashing rules.

pdyraga committed:
Fixed the value of the contextString

The values was not matching the one explained in the comment above. Now
they do match.

pdyraga committed:
Removed references to HASH-TO-CURVE

We are not using the implementation from HASH-TO-CURVE. We use BIP-340
version of hashing functions.

pdyraga commented:
Note that this specification is specific to BIP340, in the next PR I am planning to abstract out an interface that would allow us to provide another hashing strategy.
